src/nupic/algorithms/ColumnPooler.cpp

```diff
@@ -0,0 +1,565 @@
+/* ---------------------------------------------------------------------
+ * Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2013, Numenta, Inc.
+ * Copyright (C) 2019, David McDougall
+ *
+ * Unless you have an agreement with Numenta, Inc., for a separate license for
+ * this software code, the following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ----------------------------------------------------------------------
+ */
+
+/** @file
+ * Implementation of ColumnPooler
+ */
+
+#include <iostream>
+#include <vector>
+#include <string>
+#include <algorithm>
+#include <iomanip> // std::setprecision
+#include <functional> // function
+
+// #include <nupic/algorithms/ColumnPooler.hpp>
+#include <nupic/types/Types.hpp>
+#include <nupic/types/Serializable.hpp>
+#include <nupic/types/Sdr.hpp>
+#include <nupic/utils/SdrMetrics.hpp>
+#include <nupic/algorithms/Connections.hpp>
+#include <nupic/algorithms/TemporalMemory.hpp>
+#include <nupic/math/Math.hpp>
+#include <nupic/math/Topology.hpp>
+
+namespace nupic {
+namespace algorithms {
+namespace column_pooler {
+
+using namespace std;
+using namespace nupic;
+using namespace nupic::math::topology;
+using namespace nupic::algorithms::connections;
+using nupic::algorithms::temporal_memory::TemporalMemory;
+
+
+// TODO: Connections learning rules are different.
+
+
+/* Topology( location, potentialPool ) */
+typedef function<void(SDR&, SDR&)> Topology_t;
+
+class ColumnPooler // : public Serializable
+{
+private:
+  vector<UInt> proximalInputDimensions_;
+  vector<UInt> distalInputDimensions_;
+  vector<UInt> inhibitionDimensions_;
+  vector<UInt> cellDimensions_;
```

Comment:
can probably all be const

Reply:
These are private and non-const.  These need to be non-constant so that the class can initialize them & work with them. 

There are references to these which are public & constant.

Reply:
this is really just a nitpick, but I assume any of these dimensions cannot change during program's runtime? 
Then those could be const and use "early initializers" ... 
```
constructor(params, ..vGivenCellDims,...) : cellDimensions_(vGivenCellDims),.. { 
//what you do in constrcutor
}
```

Reply:
I don' think that would work with the default constructor & initialize method `ColumnPooler() {}; initialize(Parameters)`

Reply:
`ColumnPooler() : const1(Parameters.const1) {}; ` 
but let's leave it be then, Serialization will have trouble with it.